docs/grammar.md

```diff
@@ -0,0 +1,84 @@
+# Grammar
+
+ABNF syntax as per [RFC5234: Augmented BNF for Syntax Specifications: ABNF](https://datatracker.ietf.org/doc/html/rfc5234).
+
+```abnf
+purl              = scheme ":" *"/" type
+                    [ 1*"/" namespace ] 1*"/" name *"/"
+                    [ "@" version ] [ "?" qualifiers ] [ "#" *"/" subpath *"/" ]
+                            ; leading and trailing slashes allowed here and there
+purl-canonical    = scheme ":"      type-canonical
+                    [   "/" namespace ]   "/" name
+                    [ "@" version ] [ "?" qualifiers ] [ "#"      subpath      ]
+
+scheme            = %x70.6B.67    ; lowercase string "pkg"
+                            ; per ABNF spec: strings are case insensitive [...] To specify a rule that is case sensitive, specify the characters individually.
+
+type              =    ALPHA *(    ALPHA / DIGIT / "." / "-" )
+type-canonical    = LOWALPHA *( LOWALPHA / DIGIT / "." / "-" )
+
+namespace         = namespace-segment *( "/" namespace-segment )
+namespace-segment = 1*namespace-sc
+namespace-sc      = PERM-ALPHANUM
+                  / PERM-PUNCTUATION
+                  / "%" ( PERM-ESCAPED-00-1F
+                        / %x32 ( DIGIT / "A" / "B" / "C" )    ; 20-2F - except seperator "/"(%2F) and general exclusion "."(%2E) and "-"(%2D)
+                        / PERM-ESCAPED-30-FF )
+                            ; namespace safe characters
```

Comment:
For different reasons, but this production is the same as the one in `subpath-sc`.
It might be useful grouping them in `PERM-ESCAPED-NO-SLASH` or similar.

Note that the word “separator” is misspelled here and a couple of other places.